process/block/preprocess/sovereignChainTransactions.go

```diff
@@ -151,9 +157,70 @@ func (sct *sovereignChainTransactions) shouldContinueProcessingScheduledTx(
 
 	addressHasEnoughBalance := sct.hasAddressEnoughInitialBalance(tx)
 	if !addressHasEnoughBalance {
+		log.Debug("address has not enough initial balance", "sender", sct.pubkeyConverter.Encode(tx.SndAddr))
 		mbInfo.schedulingInfo.numScheduledTxsWithInitialBalanceConsumed++
 		return nil, nil, false
 	}
 
 	return tx, miniBlock, true
 }
+
+func (sct *sovereignChainTransactions) shouldSkipMiniBlock(miniBlock *block.MiniBlock) bool {
+	shouldSkipMiniBlock := miniBlock.SenderShardID != sct.shardCoordinator.SelfId() ||
+		!sct.isMiniBlockCorrect(miniBlock.Type)
+
+	return shouldSkipMiniBlock
+}
+
+func (sct *sovereignChainTransactions) isVerifyTransactionFailed(tx data.TransactionHandler, senderAccount state.UserAccountHandler, err error) bool {
+	if err != nil && !errors.Is(err, process.ErrHigherNonceInTransaction) {
+		log.Debug("sovereignChainTransactions.isVerifyTransactionFailed", "error", err)
+		return true
+	}
+
+	if check.IfNil(senderAccount) {
+		log.Debug("sovereignChainTransactions.isVerifyTransactionFailed", "error", process.ErrNilUserAccount)
+		return true
+	}
+
+	sct.mutSenderInfo.Lock()
+	defer sct.mutSenderInfo.Unlock()
+
+	sndAddr := string(tx.GetSndAddr())
+	si, ok := sct.senderInfo[sndAddr]
+	if !ok {
+		si = &senderInfo{
+			nonce:   senderAccount.GetNonce(),
+			balance: big.NewInt(0).Set(senderAccount.GetBalance()),
+		}
+	}
+
+	if si.nonce < tx.GetNonce() {
+		log.Debug("sovereignChainTransactions.isVerifyTransactionFailed", "error", process.ErrHigherNonceInTransaction,
+			"account nonce", si.nonce,
+			"tx nonce", tx.GetNonce())
+		return true
+	}
+
+	txFee := sct.economicsFee.ComputeTxFee(tx)
+	if si.balance.Cmp(txFee) < 0 {
+		log.Debug("sovereignChainTransactions.isVerifyTransactionFailed", "error", process.ErrInsufficientFee,
+			"account balance", si.balance.String(),
+			"fee needed", txFee.String())
+		return true
+	}
+
+	cost := big.NewInt(0).Add(txFee, tx.GetValue())
+	if si.balance.Cmp(cost) < 0 {
+		log.Debug("sovereignChainTransactions.isVerifyTransactionFailed", "error", process.ErrInsufficientFunds,
+			"account balance", si.balance.String(),
+			"cost", cost.String())
+		return true
+	}
+
+	si.nonce++
+	si.balance = big.NewInt(0).Sub(si.balance, cost)
```

Comment:
It is enough to do si.balance.Sub(si.balance, cost)

Reply:
Done